frontend/src/components/widgets/DataFrame/columns/utils.ts

```diff
@@ -0,0 +1,358 @@
+ */
+interface MissingValueCell extends TextCell {
+  readonly isMissingValue: true
+}
+
+/**
+ * Returns `true` if the given cell contains no value (-> missing value).
+ * For example, a number cell that contains null is interpreted as a missing value.
+ */
+export function isMissingValueCell(cell: GridCell): cell is MissingValueCell {
+  return (
+    cell.hasOwnProperty("isMissingValue") &&
+    (cell as MissingValueCell).isMissingValue
+  )
+}
+
+/**
+ * Returns an empty cell.
+ */
+export function getEmptyCell(): LoadingCell {
+  return {
+    kind: GridCellKind.Loading,
+    allowOverlay: false,
+  } as LoadingCell
+}
+
+/**
+ * Returns an empty text cell.
+ *
+ * @param readonly: If true, returns a read-only version of the cell.
+ * @param faded: If true, returns a faded version of the cell.
+ *
+ * @return a GridCell object that can be used by glide-data-grid.
+ */
+export function getTextCell(readonly: boolean, faded: boolean): TextCell {
+  const style = faded ? "faded" : "normal"
+  return {
+    kind: GridCellKind.Text,
+    data: "",
+    displayData: "",
+    allowOverlay: true,
+    readonly,
+    style,
+  } as TextCell
+}
+
+/**
+ * Converts from our BaseColumn format to the glide-data-grid compatible GridColumn.
+ */
+export function toGlideColumn(column: BaseColumn): GridColumn {
+  return {
+    id: column.id,
+    title: column.title,
+    hasMenu: false,
+    themeOverride: column.themeOverride,
+    ...(column.isStretched && {
+      grow: column.isIndex ? 1 : 3,
+    }),
+    ...(column.width && {
+      width: column.width,
+    }),
+  } as GridColumn
+}
+
+/**
+ * Merges the default column parameters with the user-defined column parameters.
+ *
+ * @param defaultParams - The default column parameters.
+ * @param userParams - The user-defined column parameters.
+ *
+ * @returns The merged column parameters.
+ */
+export function mergeColumnParameters(
+  defaultParams: Record<string, any> | undefined | null,
+  userParams: Record<string, any> | undefined | null
+): Record<string, any> {
+  if (isNullOrUndefined(defaultParams)) {
+    return userParams || {}
+  }
+
+  if (isNullOrUndefined(userParams)) {
+    return defaultParams || {}
+  }
+
+  return merge(defaultParams, userParams)
+}
+
+/**
+ * Converts the given value of unknown type to an array without
+ * the risks of any exceptions.
+ *
+ * @param data - The value to convert to an array.
+ *
+ * @returns The converted array or an empty array if the value cannot be interpreted as an array.
+ */
+export function toSafeArray(data: any): any[] {
+  if (isNullOrUndefined(data)) {
+    return []
+  }
+
+  if (typeof data === "number" || typeof data === "boolean") {
+    // Single number or boolean
+    return [data]
+  }
+
+  if (typeof data === "string") {
+    if (data === "") {
+      // Empty string
+      return []
+    }
+
+    // Try to parse string to an array
+    if (data.trim().startsWith("[") && data.trim().endsWith("]")) {
```

Comment:
Should this method support list of lists and do some recursive searching? I could maybe imagine a world of getting a list of lists into a ListColumn. I looked later down into the method and not sure if there is a catch for that.

Reply:
The current `ListColumn` cannot really deal with lists in lists, and it would take too much effort, for now, to make it compatible. But eventually, we should support more complex nested types (also dicts). We could have something like `st.json` for viewing and maybe event editing this data structures. So, it's something for the future roadmap.